Implement support for breakpoint events

Phantomical opened the pull request:
This PR adds support for using hardware breakpoint events. 

The detailed changes are:
- `Event::r#type()` and `Event::config()` have been replaced with a single `Event::update_attrs` method which allows the event to set whichever fields it needs to, not just `type_` and `config`.
- I've added a new `Breakpoint` event type along with appropriate docs.
- I've also added some test cases that test both read/write and execute breakpoints on the current process, I have verified that these work locally.

This also adds `bitflags` as a dependency. If you don't want that I can try to replace it with some hand-written equivalent.

Phantomical committed:
Combine `Event::r#type` and `Event::config` into `Event::update_attrs`

The previous split worked fine when the only fields that were worth
setting were `type` and `config`. However, once we get to other event
types like probes and breakpoints it will be necessary to set other
fields. Having a single method to do all of that works best.

Phantomical committed:
Add support for breakpoint events

Phantomical committed:
Add test cases that test both data and executable breakpoints

Phantomical committed:
Properly document `Breakpoint`

Phantomical committed:
Tweak doc wording

jimblandy commented on perf-event/src/events.rs:
```diff
@@ -47,22 +52,34 @@ pub enum Event {
-    }
-
-    pub(crate) fn config(self) -> u64 {
+    pub(crate) fn update_attrs(self, attr: &mut bindings::perf_event_attr) {
```
This isn't a bad change, but I wonder if it should go even further. Maybe we should ditch the idea of an `Event` *value* at all, and simply have `hardware`, `software`, `cache`, and `breakpoint` methods directly on `Builder`. Whatever fields the `Event` enum variants had would become arguments to those methods.

What's important to preserve here is that you can't select an event kind without providing all the necessary information to get a meaningful result - like, say, a breakpoint without an address. If we just let people set `perf_event_attr` fields directly, that would be a risk. But by making people produce an `Event` the idea was that you'd make sure they'd filled everything in.

Replacing each `Event` variant with a method on `Builder` could accomplish the same goal, and remove the pointless indirection of "first we build the `Event`; then we wait for them to hand it to us (which is the only thing you can do with an `Event` anyway); then we move the fields into the `perf_event_attr` struct." Instead, you'd just have the one `Builder` method that just does the deed.

But I don't think that needs to be done as part of this PR unless you feel like it. Filed as #23.

jimblandy commented on perf-event/src/events.rs:
```diff
@@ -318,3 +335,172 @@ pub enum CacheResult {
+///
+/// [man]: http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Breakpoint {
```
I think we can tighten this up without making it any harder to use:

```
pub enum Breakpoint {
    Data { access: BreakpointAccess, addr: u64, len: u64 },
    Code { addr: u64 },
}

enum BreakpointAccess { Read, Write, ReadWrite }
```

This conveys the underlying API's constraints, but still seems straightforward to use. It could have an `update_attrs` method for `Event::update_attrs` to call.

We could use an enum for the length, too, but that just seems like going too far.

Phantomical replied:
That is nicer. I've gone and adjusted `Breakpoint` to match although I've left `BreakpointAccess` as a bitfield since it matches what is going on with the underlying `perf_event_open` API better.

jimblandy commented on perf-event/src/events.rs:
```diff
@@ -318,3 +335,172 @@ pub enum CacheResult {
+            len: std::mem::size_of::<libc::c_long>() as _,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we read from the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn read(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::READ,
+            addr,
+            len,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we write to the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn write(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::WRITE,
+            addr,
+            len,
+        }
+    }
+
+    /// Create a breakpoint configuration to count the number of times that
+    /// we either read from or write to the provided memory location.
+    ///
+    /// See the struct field docs for valid values of `len`.
+    pub const fn read_write(addr: u64, len: u64) -> Self {
+        Self {
+            ty: BreakpointType::READ_WRITE,
+            addr,
+            len,
+        }
+    }
+}
```
All this could go away - we'd just let people construct `Breakpoint` variants.

Phantomical replied:
I've left these in for now since I think that, as a user of the API, they make it easier to use. If you really want them gone then I'll remove them.

jimblandy commented on perf-event/src/events.rs:
```diff
@@ -318,3 +335,172 @@ pub enum CacheResult {
+    }
+}
+
+/// A hardware breakpoint.
```
Great docs. I really appreciate this.

jimblandy commented on perf-event/tests/breakpoint.rs:
```diff
@@ -0,0 +1,64 @@
+        .expect("Unable to build performance counter");
+    counter.enable().unwrap();
+
+    for _ in 0..1000 {
+        copy_and_fwrite(&data, &mut file).unwrap();
+    }
```
For the data test, at least, let's have a function that uses `read_volatile` on a raw pointer. The compiler is (technically) free to change this to perform a single read to copy the data elsewhere and then do all the writes from the copy. But volatile reads are intended to be used when the access provokes some sort of hardware interaction - which seems like a pretty good fit for this test.

Phantomical replied:
I had actually forgotten that `read_volatile` existed. I've updated it and verified that the test works on my machine in both debug and release mode.

jimblandy commented on perf-event/tests/breakpoint.rs:
```diff
@@ -0,0 +1,64 @@
+        .expect("Unable to build performance counter");
+    counter.enable().unwrap();
+
+    for _ in 0..1000 {
+        copy_and_fwrite(&data, &mut file).unwrap();
+    }
```
Isn't `#[inline_never]` and taking the function's address enough to ensure that these calls will actually happen? If we haven't actually observed problems, let's ditch the I/O.

jimblandy reviewed (changes requested):
This looks fantastic. I really appreciate the care you put into the docs and tests.

There are some changes I'd like.

Phantomical committed:
Change event::Breakpoint to match review suggestions

Phantomical committed:
Remove file IO from test cases

Phantomical committed:
Formatting

jimblandy committed:
Make Breakpoint fn docs read better in search summaries.

All these descriptions began with "Create a breakpoint configuration
to...", but rustdoc searches show the beginning of each hit's first
doc paragraph, so it's nice to get the most distinctive information
about the function as early in the leading sentence as you can.

jimblandy commented on perf-event/src/events.rs:
```diff
@@ -47,22 +52,46 @@ pub enum Event {
+                match bp {
+                    Breakpoint::Data { access, addr, len } => {
+                        attr.bp_type = access.bits();
+                        attr.__bindgen_anon_3.bp_addr = addr;
+                        attr.__bindgen_anon_4.bp_len = len;
```
Wow, this really is the only way to set that field. That seems really fragile, although I'm not sure what else bindgen could do with a transparent union.

jimblandy replied:
Well... obviously, it should generate a setter on the enclosing struct.

jimblandy reviewed (approved):
Thank you!
